dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -692,48 +692,83 @@ DecodeDetail ExchangeReceiverBase<RPCContext>::decodeChunks(
     {
-        Block block = CHBlockChunkCodec::decode(*chunk, header);
-        detail.rows += block.rows();
-        if (unlikely(block.rows() == 0))
+        auto result = decoder_ptr->decodeAndSquash(*chunk);
+        if (!result)
             continue;
-        block_queue.push(std::move(block));
+        detail.rows += result->rows();
+        if likely (result->rows() > 0)
+        {
+            block_queue.push(std::move(result.value()));
+        }
     }
     return detail;
 }
 
 template <typename RPCContext>
-ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::nextResult(std::queue<Block> & block_queue, const Block & header, size_t stream_id)
+ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::nextResult(
+    std::queue<Block> & block_queue,
+    const Block & header,
+    size_t stream_id,
+    std::unique_ptr<IChunkDecodeAndSquash> & decoder_ptr)
 {
     if (unlikely(stream_id >= msg_channels.size()))
     {
         LOG_ERROR(exc_log, "stream_id out of range, stream_id: {}, total_stream_count: {}", stream_id, msg_channels.size());
         return ExchangeReceiverResult::newError(0, "", "stream_id out of range");
     }
+
     std::shared_ptr<ReceivedMessage> recv_msg;
     if (msg_channels[stream_id]->pop(recv_msg) != MPMCQueueResult::OK)
     {
-        std::unique_lock lock(mu);
-        return state != ExchangeReceiverState::NORMAL
-            ? ExchangeReceiverResult::newError(0, name, constructStatusString(state, err_msg))
-            : ExchangeReceiverResult::newEOF(name); /// live_connections == 0, msg_channel is finished, and state is NORMAL, that is the end.
+        return handleUnnormalChannel(block_queue, decoder_ptr);
     }
     else
     {
         assert(recv_msg != nullptr);
         if (unlikely(recv_msg->error_ptr != nullptr))
             return ExchangeReceiverResult::newError(recv_msg->source_index, recv_msg->req_info, recv_msg->error_ptr->msg());
-        return toDecodeResult(block_queue, header, recv_msg);
+        return toDecodeResult(block_queue, header, recv_msg, decoder_ptr);
+    }
+}
+
+template <typename RPCContext>
+ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::handleUnnormalChannel(
+    std::queue<Block> & block_queue,
+    std::unique_ptr<IChunkDecodeAndSquash> & decoder_ptr)
+{
+    std::optional<Block> last_block;
+    if (decoder_ptr)
+        last_block = decoder_ptr->flush();
+
+    std::unique_lock lock(mu);
+    if (this->state != DB::ExchangeReceiverState::NORMAL)
+    {
+        return DB::ExchangeReceiverResult::newError(0, DB::ExchangeReceiverBase<RPCContext>::name, DB::constructStatusString(this->state, this->err_msg));
+    }
+    else
+    {
+        /// If there are cached data in squashDecoder, then just push the block and return EOF next iteration
+        if (last_block && last_block->rows() > 0)
+        {
+            const auto & result = DB::ExchangeReceiverResult::newRowsInfoOnly(last_block->rows());
```

Comment:
Why not return `ExchangeReceiverResult::newOk()`?
What is the difference between `newOk` and `newRowsInfoOnly`?

Reply:
Introduce a new flag to indicate that the call_index and other info in the result is not usable.

Reply:
Actually, it is dirty here, since the squash logic will seperate block and packets, somehow breaks the outside expection.